Feature: accessing versioned resource via GET endpoints using URIs

cmark opened the pull request:
This PR adds the ability to access a versioned form of a resource via the following endpoints:
* `GET /bundles/:id`
* `GET /codesystems/:id`

The id argument now accepts a complete URI, including both version paths and timestamp parts.
Examples:
* `GET /codesystems/SNOMEDCT/2022-07-31`
* `GET /codesystems/SNOMEDCT@1659225600`

cmark committed:
feat(api): support versioned resource access in GET endpoints

Accessing both via versionId path or URI with timestamp fragment.

cmark committed:
feat(api): support subpath parsing in `GET /bundles/:id` endpoint

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/bundle/Bundle.java:
```diff
@@ -89,6 +90,10 @@ public Builder toDocumentBuilder() {
 	}
+
+	public static ResourceURI uri(String bundleUri) {
+		return bundleUri.startsWith(RESOURCE_TYPE) ? new ResourceURI(bundleUri) : ResourceURI.of(RESOURCE_TYPE, bundleUri);
```
We might want to add a trailing `/` to the constant in these `startsWith` checks. As an example, `"bundlesort"` would be a resource-type-less URI, but the code treats it as one that has a type because it starts with `"bundles"`.

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/internal/ResourceRepository.java:
```diff
@@ -140,10 +141,10 @@ public void run(StagingArea staging) {
-					.execute(context)
+					.stream(context)
+					.flatMap(Versions::stream)
 					.forEach(version -> {
```
As this became a `Stream`, `forEachOrdered` should be used so we don't get caught surprised by a future version of Java that enables wider parallel processing.

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/resource/BaseGetResourceRequest.java:
```diff
@@ -0,0 +1,66 @@
+			return ResourceRequests.prepareSearchVersion()
+					.one()
+					.filterByResource(resourceUri.withoutPath())
+					.filterByVersionId(resourceUri.getPath())
```
We might be able to come up with a filter in the future that matches any descendant of a version branch path. Bookmarking! 🔖

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/resource/BaseGetResourceRequestBuilder.java:
```diff
@@ -0,0 +1,36 @@
+	B extends GetResourceRequestBuilder<B, SB, RepositoryContext, SR, R>, 
+	SB extends SearchResourceRequestBuilder<SB, RepositoryContext, SR>, 
+	SR, 
+	R> extends GetResourceRequestBuilder<B, SB, RepositoryContext, SR, R> implements ResourceRepositoryRequestBuilder<R> {
```
😖 Please move this last `extends` to its own line, so there is a separation between the type parameters and the base class of this builder.

apeteri reviewed (approved):
👍

cmark committed:
fix(api): support accessing the version denoted by the LATEST keyword

cmark committed:
fix(api): use dedicated endpoint to access versioned resource...

...metadata instead of allowing the use of subpaths

Path regex unfortunately clashes with other dedicated routes and will
swallow requests supposed to be routed elsewhere.

cmark committed:
fix(api): use resourceId only when searching for versions

For now, fall back to the HEAD version when there is no version found on
the selected path.
Fix Bundle API initialization.

cmark committed:
fix(test): incorrect assertion (HTTP 400 vs 404) when path is not found

cmark committed:
fix(api): ensure we always have a default `en` languageTag set...

...in the Accept-Language header default value

cmark committed:
fix(api): append slash when checking whether a string representation...

...of an URI contains the type already or not

cmark committed:
fix(core): use forEachOrdered in ResourceRepository

cmark committed:
chore: add new line before extends statement

nagyo reviewed (approved):
semi-RSLGTM!
